cmd/backendapi/apisrv/apisrv.go

```diff
@@ -115,112 +119,144 @@ func (s *backendAPI) CreateMatch(c context.Context, p *backend.Profile) (*backen
 			"component": "statestorage",
-		}).Error("Statestorage failure to queue profile")
+		}).Error("State storage failure to queue profile")
 
 		// Failure! Return empty match object and the error
 		stats.Record(fnCtx, BeGrpcErrors.M(1))
 		return &backend.MatchObject{}, err
 	}
+	beLog.Info("Profile added to processing queue")
 
-	beLog.WithFields(log.Fields{
-		"profileID":         p.Id,
-		"matchObjectID":     moID,
-		"profileRequestKey": profileRequestKey,
-	}).Info("Profile added to processing queue")
-
-	// get and return matchobject
-	watchChan := redisHelpers.Watcher(ctx, s.pool, profileRequestKey) // Watcher() runs the appropriate Redis commands.
-	mo := &backend.MatchObject{Id: p.Id, Properties: ""}
-	errString := ("Error retrieving matchmaking results from statestorage")
+	// get and return matchobject, it will be written to the requestKey when the MMF has finished.
+	var ok bool
+	newMO := backend.MatchObject{Id: requestKey}
+	watchChan := redispb.Watcher(ctx, s.pool, newMO) // Watcher() runs the appropriate Redis commands.
+	errString := ("Error retrieving matchmaking results from state storage")
 	timeout := time.Duration(s.cfg.GetInt("interval.resultsTimeout")) * time.Second
 
 	select {
 	case <-time.After(timeout):
 		// TODO:Timeout: deal with the fallout.  There are some edge cases here.
 		// When there is a timeout, need to send a stop to the watch channel.
 		stats.Record(fnCtx, BeGrpcRequests.M(1))
-		return mo, errors.New(errString + ": timeout exceeded")
+		return profile, errors.New(errString + ": timeout exceeded")
 
-	case properties, ok := <-watchChan:
+	case newMO, ok = <-watchChan:
 		if !ok {
-			// ok is false if watchChan has been closed by redisHelpers.Watcher()
-			stats.Record(fnCtx, BeGrpcRequests.M(1))
-			return mo, errors.New(errString + ": channel closed - was the context cancelled?")
-		}
-
-		beLog.WithFields(log.Fields{
-			"profileRequestKey": profileRequestKey,
-			"matchObjectID":     moID,
-			// DEBUG ONLY: This prints the entire result from redis to the logs
-			"matchProperties": properties, // very verbose!
-		}).Debug("Received match object from statestorage")
-
-		// 'ok' was true, so properties should contain the results from redis.
-		// Do some error checking on the returned JSON
-		if !gjson.Valid(properties) {
-			// Just splitting this across lines for readability/wrappability
-			thisError := ": Retreived json was malformed"
-			thisError = thisError + " - did the evaluator write a valid JSON match object?"
-			stats.Record(fnCtx, BeGrpcErrors.M(1))
-			return mo, errors.New(errString + thisError)
+			// ok is false if watchChan has been closed by redispb.Watcher()
+			newMO.Error = newMO.Error + "; channel closed - was the context cancelled?"
+		} else {
+			// 'ok' was true, so properties should contain the results from redis.
+			// Do basic error checking on the returned JSON
+			if !gjson.Valid(profile.Properties) {
+				newMO.Error = "retreived properties json was malformed"
```

Comment:
The matchobject error gets override here. Is there a grpc oriented way of throwing an error instead?

Reply:
Could also do the append like above.

Reply:
Hrm.  At that point maybe better to make Error a repeated string message field.  I'd rather people be able to 'loop' through the errors, trying to identify an atomic string that has a specific failure meaning than having to do do some kind of fuzzy/substring matching within a concatenated string. This could also allow us to raise errors from further down the stack? 
 Thoughts?

Reply:
Is errors reserved for open match errors or is it available to match functions?
Either way, having errors as a repeated string probably makes more sense. I know in C# There's innerexceptions, but i don't think we need nearly that much wrapping here to get the errors communicated.